knowhere/index/vector_index/IndexDiskANN.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#include "index/vector_index/IndexDiskANN.h"
+
+#include <omp.h>
+
+#include <limits>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "DiskANN/include/aux_utils.h"
+#ifndef _WINDOWS
+#include "DiskANN/include/linux_aligned_file_reader.h"
+#else
+#include "DiskANN/include/windows_aligned_file_reader.h"
+#endif
+#include "knowhere/common/Exception.h"
+#include "knowhere/index/vector_index/IndexDiskANNConfig.h"
+#include "knowhere/index/vector_index/adapter/VectorAdapter.h"
+
+namespace knowhere {
+
+template <typename T>
+IndexDiskANN<T>::IndexDiskANN(std::string index_prefix, MetricType metric_type,
+                              std::unique_ptr<FileManager> file_manager)
+    : index_prefix_(index_prefix), file_manager_(std::move(file_manager)) {
+    index_type_ = IndexEnum::INDEX_DISKANN;
+
+    if (metric_type == metric::L2) {
+        metric_ = diskann::L2;
+    } else if (metric_type == metric::IP) {
+        if (!std::is_same_v<T, float>) {
+            KNOWHERE_THROW_MSG(
+                "DiskANN currently only supports floating point data for Max "
+                "Inner Product Search. ");
+        }
+        metric_ = diskann::INNER_PRODUCT;
+    } else {
+        KNOWHERE_THROW_MSG("DiskANN only support L2 and IP distance.");
+    }
+}
+
+namespace {
+void
+CheckPreparation(bool is_prepared) {
+    if (is_prepared) {
+        KNOWHERE_THROW_MSG("DiskANN is not prepared yet, plz call Prepare() to make it ready for queries.");
+    }
+}
+
+/**
+ * @brief Convert id from uint64 to int64. We can avoid this by either supporting int64 id in DiskANN or accepting
+ * uint64 id in Milvus.
+ *
+ * @return false if any error.
+ */
+auto
+ConvertId(const uint64_t* src, int64_t* des, size_t num) -> bool {
+    int64_t max = std::numeric_limits<int64_t>::max();
+    bool successful = true;
+#pragma omp parallel for schedule(static, 65536)
+    for (size_t n = 0; n < num; ++n) {
+        if (src[n] > max) {
```

Comment:
this compare will convert src[n] to int64_t implicitly, so it will always be false

Reply:
I tried this locally
```
int main() {
        int64_t max =  numeric_limits<int64_t>::max();
        uint64_t* src = static_cast<uint64_t*>(malloc(sizeof(uint64_t) ));
        *src = 9223372036854775809;
        if (src[0] > max) {
                cout << "yes" << endl;
        }

}
```

And it printed "yes"

Reply:
this way is not safe in all platforms, better use this:
        if (src[n] && (1 << 63)) {